CommonTools/ParticleFlow/plugins/PFCandidateRecalibrator.cc

```diff
@@ -0,0 +1,326 @@
+		
+		bool toKill = false;
+			
+		for(auto const& badIt: badChHF_)
+		  {
+		    if ( hDetId.ieta() == badIt.ieta &&
+			 hDetId.iphi() == badIt.iphi )
+		      {
+			LogDebug("PFCandidateRecalibrator") << "==> orig en (tot,H,E): " 
+							    << pf.energy() << " " << pf.rawHcalEnergy() << " " << pf.rawEcalEnergy();
+			if(badIt.depth == 1) //depth1
+			  {
+			    longE *= badIt.ratio;
+			    ecalEnergy = longE - shortE;
+			    totEnergy = ecalEnergy + hcalEnergy;
+			  }
+			else //depth2
+			  {
+			    shortE *= badIt.ratio;
+			    hcalEnergy = 2*shortE;
+			    if(ecalEnergy > 0)
+			      ecalEnergy = longE - shortE;
+			    totEnergy = ecalEnergy + hcalEnergy;
+			  }
+			//kill candidate if goes below thr
+			if((pf.particleId() == reco::PFCandidate::ParticleType::h_HF && shortE < 1.4) || 
+			   (pf.particleId() == reco::PFCandidate::ParticleType::egamma_HF && longE < 1.4))
+			  toKill = true;
+
+			LogDebug("PFCandidateRecalibrator") << "====> ieta,iphi,depth: " 
+							    << badIt.ieta << " " << badIt.iphi << " " << badIt.depth << " corr: " << badIt.ratio;
+			LogDebug("PFCandidateRecalibrator") << "====> recal en (tot,H,E): " 
+							    << totEnergy << " " << hcalEnergy << " " << ecalEnergy;
+
+		      }
+		  }
+		
+		if(toKill == true)
+		  {
+		    discarded->push_back(pf);
+		    oldToNew[i] = (-discarded->size());
+		    badToOld.push_back(i);
+
+		    LogDebug("PFCandidateRecalibrator") << "==> KILLED ";
+		  }
+		else
+		  {
+		    copy->push_back(pf);
+		    oldToNew[i] = (copy->size());
+		    newToOld.push_back(i);
+		    
+		    copy->back().setHcalEnergy(hcalEnergy, hcalEnergy);
+		    copy->back().setEcalEnergy(ecalEnergy, ecalEnergy);
+		    math::XYZTLorentzVector recalibP4(pf.px()*(totEnergy/totEnergyOrig), pf.py()*(totEnergy/totEnergyOrig), pf.pz()*(totEnergy/totEnergyOrig), totEnergy);
+		    copy->back().setP4( recalibP4 );
+
+		    LogDebug("PFCandidateRecalibrator") << "====> stored en (tot,H,E): " 
+							<< copy->back().energy() << " " << copy->back().hcalEnergy() << " " << copy->back().ecalEnergy();
+		  }
+	      }
+	    else
+	      {
+		copy->push_back(pf);
+		oldToNew[i] = (copy->size());
+		newToOld.push_back(i);
+	      }
+	  }
+	else
+	  {
+	    copy->push_back(pf);
+	    oldToNew[i] = (copy->size());
+	    newToOld.push_back(i);
+	  }
+      }
+
+    // Now we put things in the event
+    edm::OrphanHandle<reco::PFCandidateCollection> newpf = iEvent.put(std::move(copy));
+    edm::OrphanHandle<reco::PFCandidateCollection> badpf = iEvent.put(std::move(discarded), "discarded");
+
+    std::unique_ptr<edm::ValueMap<reco::PFCandidateRef>> pf2pf(new edm::ValueMap<reco::PFCandidateRef>());
+    edm::ValueMap<reco::PFCandidateRef>::Filler filler(*pf2pf);
+    std::vector<reco::PFCandidateRef> refs; refs.reserve(nPfCand);
+
+    // old to new
+    for (i = 0; i < nPfCand; ++i) {
+      if (oldToNew[i] > 0) {
+	refs.push_back(reco::PFCandidateRef(newpf, oldToNew[i]-1));
```

Comment:
copy from https://github.com/cms-sw/cmssw/pull/22941#discussion_r182130902

> this can be a for range loop ```for (auto iOldToNew : oldToNew){ ... ```

Reply:
agreed, will be there at the next commit